game_mechanics.py

```diff
@@ -1,71 +1,449 @@
-    pass
+HERE = Path(__file__).parent.resolve()
 
 
-def load_pkl(team_name: str):
-    """Load a user PKL."""
-    pass
+def load_network(team_name: str) -> nn.Module:
+    net_path = os.path.join(HERE, f"{team_name}_network.pt")
+    assert os.path.exists(
+        net_path
+    ), f"Network saved using TEAM_NAME='{team_name}' doesn't exist! ({net_path})"
+    model = torch.load(net_path)
+    model.eval()
+    return model
 
 
-def choose_move_randomly(state):
-    pass
+def save_network(network: nn.Module, team_name: str) -> None:
+    assert isinstance(
+        network, nn.Module
+    ), f"train() function outputs an network type: {type(network)}"
+    assert "/" not in team_name, "Invalid TEAM_NAME. '/' are illegal in TEAM_NAME"
+    net_path = HERE / f"{team_name}_network.pt"
+    n_retries = 5
+    for attempt in range(n_retries):
+        try:
+            torch.save(network, net_path)
+            load_network(team_name)
+            return
+        except Exception:
+            if attempt == n_retries - 1:
+                raise
 
 
-def play_game(
-    your_choose_move: Callable,
-    opponent_choose_move: Callable,
-    game_speed_multiplier=1,
-    render=True,
-    verbose=False,
-):
-    pass
+########## POTENTIALLY USEFUL FEATURES ############################
 
 
-class Env:
-    def __init__(self):
-        """Initialises the object.
+def get_legal_moves(board: np.ndarray) -> List[Tuple[int, int]]:
+    """Return a list of legal moves that can be made by player 1 on the current board."""
+    return _get_legal_moves(board, 1)
 
-        Is called when you call `environment = Env()`.
 
-        It sets everything up in the starting state for the episode to run.
-        """
+def make_move(board: np.ndarray, move: Tuple[int, int]) -> np.ndarray:
+    """Returns board after move has been made by player 1."""
+    return _make_move(board, move, 1)
+
+
+def choose_move_randomly(
+    board: np.ndarray,
+) -> Optional[Tuple[int, int]]:
+    """Returns a random legal move on the current board (always plays as player 1."""
+    moves = get_legal_moves(board)
+    if moves:
+        return random.choice(moves)
+    return None
+
+
+####### THESE FUNCTIONS ARE LESS USEFUL ############
+
+
+def _make_move(board: np.ndarray, move: Tuple[int, int], current_player: int) -> np.ndarray:
+    """Returns board after move has been made."""
+    if is_legal_move(board, move, current_player):
+        board_after_move = copy.deepcopy(board)
+        board_after_move[move] = current_player
+        board_after_move = flip_tiles(board_after_move, move, current_player)
+    else:
+        raise ValueError(f"Move {move} is not a valid move!")
+    return board_after_move
+
+
+def is_legal_move(board: np.ndarray, move: Tuple[int, int], current_player: int) -> bool:
+    """Method: is_legal_move
+    Parameters: self, move (tuple)
+    Returns: boolean (True if move is legal, False otherwise)
+    Does: Checks whether the player's move is legal.
```

Comment:
Strange format & out of date

Aren't we using Google docstrings?

Reply:
yeah it's called the "Copy and pasted from the internet" formar